#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
gRPC server for bazi-core-service.
"""

from __future__ import annotations

import json
import os
import sys
from concurrent import futures

import grpc

BASE_DIR = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
PROJECT_ROOT = BASE_DIR
sys.path.insert(0, PROJECT_ROOT)

# å¯¼å…¥ç”Ÿæˆçš„ gRPC ä»£ç 
sys.path.insert(0, os.path.join(PROJECT_ROOT, "proto", "generated"))
import bazi_core_pb2
import bazi_core_pb2_grpc

from src.bazi_core.calculator import BaziCoreCalculator


class BaziCoreServicer(bazi_core_pb2_grpc.BaziCoreServiceServicer):
    """å®ç° BaziCoreService çš„ gRPC æœåŠ¡"""

    def CalculateBazi(self, request: bazi_core_pb2.BaziCoreRequest, context: grpc.ServicerContext) -> bazi_core_pb2.BaziCoreResponse:
        """è®¡ç®—å…«å­—æ’ç›˜"""
        import datetime
        request_time = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        print(f"[{request_time}] ğŸ“¥ bazi-core-service: æ”¶åˆ°è¯·æ±‚ - solar_date={request.solar_date}, solar_time={request.solar_time}, gender={request.gender}", flush=True)
        
        try:
            calculator = BaziCoreCalculator(
                solar_date=request.solar_date,
                solar_time=request.solar_time,
                gender=request.gender,
            )
            result = calculator.calculate()
            print(f"[{request_time}] âœ… bazi-core-service: è®¡ç®—å®Œæˆ", flush=True)
            
            if result is None:
                context.set_code(grpc.StatusCode.INTERNAL)
                context.set_details("å…«å­—æ’ç›˜å¤±è´¥")
                return bazi_core_pb2.BaziCoreResponse()

            # è½¬æ¢ç»“æœä¸º protobuf æ ¼å¼
            response = bazi_core_pb2.BaziCoreResponse()
            
            # åŸºæœ¬ä¿¡æ¯ï¼ˆéœ€è¦ç‰¹æ®Šå¤„ç† lunar_dateï¼Œå®ƒæ˜¯å­—å…¸ï¼‰
            if "basic_info" in result:
                basic_info = result["basic_info"]
                for key, value in basic_info.items():
                    if key == "lunar_date" and isinstance(value, dict):
                        # lunar_date æ˜¯å­—å…¸ï¼Œéœ€è¦åºåˆ—åŒ–ä¸º JSON
                        response.basic_info[key] = json.dumps(value, ensure_ascii=False)
                    else:
                        response.basic_info[key] = str(value)
            
            # å››æŸ±ä¿¡æ¯
            if "bazi_pillars" in result:
                pillars = result["bazi_pillars"]
                for pillar_name, pillar_data in pillars.items():
                    pillar = bazi_core_pb2.Pillar(
                        stem=pillar_data.get("stem", ""),
                        branch=pillar_data.get("branch", "")
                    )
                    if pillar_name == "year":
                        response.year_pillar.CopyFrom(pillar)
                    elif pillar_name == "month":
                        response.month_pillar.CopyFrom(pillar)
                    elif pillar_name == "day":
                        response.day_pillar.CopyFrom(pillar)
                    elif pillar_name == "hour":
                        response.hour_pillar.CopyFrom(pillar)
            
            # å››æŸ±è¯¦æƒ…
            if "details" in result:
                for pillar_name, detail_data in result["details"].items():
                    detail = bazi_core_pb2.PillarDetail(
                        main_star=detail_data.get("main_star", ""),
                        nayin=detail_data.get("nayin", ""),
                        kongwang=detail_data.get("kongwang", ""),
                    )
                    if "deities" in detail_data:
                        detail.deities.extend(detail_data["deities"])
                    response.details[pillar_name].CopyFrom(detail)
            
            # åç¥ç»Ÿè®¡
            if "ten_gods_stats" in result:
                for key, value in result["ten_gods_stats"].items():
                    response.ten_gods_stats[key] = str(value)
            
            # äº”è¡Œä¿¡æ¯
            if "elements" in result:
                for key, value in result["elements"].items():
                    # ä½¿ç”¨ json.dumps è€Œä¸æ˜¯ str()ï¼Œç¡®ä¿æ ¼å¼æ­£ç¡®
                    response.elements[key] = json.dumps(value, ensure_ascii=False) if isinstance(value, dict) else str(value)
            
            # äº”è¡Œè®¡æ•°
            if "element_counts" in result:
                for key, value in result["element_counts"].items():
                    response.element_counts[key] = int(value)
            
            # å…³ç³»ä¿¡æ¯
            if "relationships" in result:
                for key, value in result["relationships"].items():
                    response.relationships[key] = str(value)
            
            # å…ƒæ•°æ®ï¼ˆJSON å­—ç¬¦ä¸²ï¼‰
            metadata = {
                "service": "bazi-core-service",
                "version": "1.0.0",
            }
            response.metadata_json = json.dumps(metadata, ensure_ascii=False)
            
            print(f"[{request_time}] âœ… bazi-core-service: å“åº”å·²è¿”å›", flush=True)
            return response
            
        except Exception as e:
            import traceback
            error_msg = f"è®¡ç®—å¤±è´¥: {str(e)}\n{traceback.format_exc()}"
            print(f"[{request_time}] âŒ bazi-core-service: é”™è¯¯ - {error_msg}", flush=True)
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"è®¡ç®—å¤±è´¥: {str(e)}")
            return bazi_core_pb2.BaziCoreResponse()

    def HealthCheck(self, request: bazi_core_pb2.HealthCheckRequest, context: grpc.ServicerContext) -> bazi_core_pb2.HealthCheckResponse:
        """å¥åº·æ£€æŸ¥"""
        return bazi_core_pb2.HealthCheckResponse(status="ok")


def serve(port: int = 9001):
    """å¯åŠ¨ gRPC æœåŠ¡å™¨"""
    # å¢åŠ çº¿ç¨‹æ± å¤§å°ä»¥å¤„ç†æ›´å¤šå¹¶å‘è¯·æ±‚
    # é…ç½® keepalive é€‰é¡¹ï¼Œé¿å… "Too many pings" é”™è¯¯
    server_options = [
        ('grpc.keepalive_time_ms', 300000),  # 5åˆ†é’Ÿ
        ('grpc.keepalive_timeout_ms', 20000),  # 20ç§’
        ('grpc.keepalive_permit_without_calls', False),
        ('grpc.http2.max_pings_without_data', 2),
        ('grpc.http2.min_time_between_pings_ms', 60000),  # 60ç§’
        ('grpc.http2.min_ping_interval_without_data_ms', 300000),  # 5åˆ†é’Ÿ
    ]
    
    server = grpc.server(
        futures.ThreadPoolExecutor(max_workers=20),
        options=server_options
    )
    bazi_core_pb2_grpc.add_BaziCoreServiceServicer_to_server(BaziCoreServicer(), server)
    
    listen_addr = f"localhost:{port}"  # ä½¿ç”¨ localhost é¿å…æƒé™é—®é¢˜
    server.add_insecure_port(listen_addr)
    
    server.start()
    print(f"âœ… Bazi Core gRPC æœåŠ¡å·²å¯åŠ¨ï¼Œç›‘å¬ç«¯å£: {port}")
    
    try:
        server.wait_for_termination()
    except KeyboardInterrupt:
        print("\n>>> æ­£åœ¨åœæ­¢æœåŠ¡...")
        server.stop(grace=5)
        print("âœ… æœåŠ¡å·²åœæ­¢")


if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser(description="å¯åŠ¨ Bazi Core gRPC æœåŠ¡")
    parser.add_argument("--port", type=int, default=9001, help="æœåŠ¡ç«¯å£ï¼ˆé»˜è®¤: 9001ï¼‰")
    args = parser.parse_args()
    serve(args.port)

